### PR TITLE
ci: testing python lib injection

### PR DESCRIPTION
**Pull request by robertomonteromiguel**
## Description
Move whole lib injection tests to system-tests.
Launch these tests for every PR.

## Checklist
- [ ] Add additional sections for `feat` and `fix` pull requests.
- [ ] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

**Review by Kyle-Verhoog — COMMENTED** [review]
Do we have a link to the corresponding system-tests PR?

**Comment by robertomonteromiguel** [comment]
> Do we have a link to the corresponding system-tests PR?

https://github.com/DataDog/system-tests/pull/629

File: .github/workflows/lib-injection.yml [hunk 1]
```diff
@@ -3,77 +3,66 @@ on:
...
+      - name: Checkout python tracer
+        uses: actions/checkout@v3
+        with:
+          path: './dd-trace-py'
+ 
+      - name: Checkout system tests
+        uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f # 2.3.4
+        with:
+          repository: 'DataDog/system-tests'
+          ref: 'robertomonteromiguel/lib_injection_integration'
```
Comment (robertomonteromiguel) [inline .github/workflows/lib-injection.yml:34]: I will remove this:

When my system-tests branch is merged to main branch
Just before merging this PR

File: .github/workflows/lib-injection.yml [hunk 2]
```diff
@@ -3,77 +3,66 @@ on:
...
-    - name: Docker Build
-      uses: docker/build-push-action@v3
-      with:
-        push: true
-        tags: ${{ steps.set_names.outputs.image_name }}
-        platforms: "linux/amd64,linux/arm64/v8"
-        context: ./lib-injection
-
-  build-and-publish-sample-app-images:
+   injection-tests-python:
```
Comment (robertomonteromiguel) [inline .github/workflows/lib-injection.yml:6]: I decided not to remove this file...but another option is moving these jobs to system-tests.yml
what do you think?

File: .github/workflows/lib-injection.yml [hunk 3]
```diff
@@ -15,65 +15,108 @@ jobs:
...
       run: |
-        DOCKER_IMAGE_TAG=${{ github.sha }}
+        
         DOCKER_IMAGE_NAME=dd-lib-python-init
         DOCKER_IMAGE_NAME=$(echo ghcr.io/${GITHUB_REPOSITORY}/${DOCKER_IMAGE_NAME} | tr '[:upper:]' '[:lower:]')
-        DOCKER_IMAGE_NAME_WITH_TAG=$(echo ${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG} | tr '[:upper:]' '[:lower:]')
+        DOCKER_IMAGE_NAME_WITH_TAG=$(echo ${DOCKER_IMAGE_NAME}:${{ github.sha }} | tr '[:upper:]' '[:lower:]')
+ 
+        #If we are building main branch, we also tag as latest_snapshot
+        if [ $(echo ${GITHUB_REF#refs/heads/} | tr / -) == 'robertomonteromiguel-lib_injection_system_tests_integration' ]; then
```
Comment (robertomonteromiguel) [inline .github/workflows/lib-injection.yml:35]: I will change to 1.x before merge

File: .github/workflows/lib-injection.yml [hunk 4]
```diff
@@ -26,66 +26,56 @@ jobs:
       id: buildx
       uses: docker/setup-buildx-action@v2
 
-    - name: Available platforms
-      run: echo ${{ steps.buildx.outputs.platforms }}
-
-    - name: Set Docker Image Tag
-      id: set_names
+    - name: Copy python tracer
```
Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:29]: I think we should keep all the `build-and-publish-init-image` job code here. This is production code that is responsible for building and publishing the container image. Having the code split between here and system-tests seems like unnecessary coupling and is particularly odd as now we'd be introducing production code to the system-tests repository.

File: .github/workflows/lib-injection.yml [hunk 5]
```diff
@@ -26,66 +26,56 @@ jobs:
...
         push: true
-        tags: ${{ steps.set_names.outputs.image_name }}
-        platforms: "linux/amd64,linux/arm64/v8"
+        tags: ${{ steps.lib-injection-tags.outputs.tag-names }}
+        platforms: 'linux/amd64,linux/arm64/v8'
         build-args: "DDTRACE_PYTHON_VERSION=${{ steps.get_version.outputs.library_version }}"
         context: ./lib-injection
 
-  build-and-publish-sample-app-images:
+  injection-tests:
```
Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:53]:
```suggestion
  test:
```

File: .github/workflows/lib-injection.yml [hunk 6]
```diff
@@ -26,66 +26,56 @@ jobs:
...
-          DOCKER_IMAGE_NAME=${{ matrix.image-name }}
-          DOCKER_IMAGE_NAME=$(echo ghcr.io/${GITHUB_REPOSITORY}/${DOCKER_IMAGE_NAME} | tr '[:upper:]' '[:lower:]')
-          DOCKER_IMAGE_NAME_WITH_TAG=$(echo ${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG} | tr '[:upper:]' '[:lower:]')
-          echo "Using image name '$DOCKER_IMAGE_NAME_WITH_TAG'"
-          echo "image_name=$DOCKER_IMAGE_NAME_WITH_TAG" >> $GITHUB_OUTPUT
-      - name: Login to Docker
-        run: docker login -u publisher -p ${{ secrets.GITHUB_TOKEN }} ghcr.io
-      - name: Docker Build
-        uses: docker/build-push-action@v3
+        lib-injection-connection: [ 'network','uds']
```
Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:62]:
```suggestion
        lib-injection-connection: ['network','uds']
```
nit

File: .github/workflows/lib-injection.yml [hunk 7]
```diff
@@ -26,66 +26,56 @@ jobs:
...
+        lib-injection-use-admission-controller: ['', 'use-admission-controller']
+        weblog-variant: [ 'dd-lib-python-init-test-django','dd-lib-python-init-test-django-gunicorn','dd-lib-python-init-test-django-uvicorn']
```
Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:64]:
```suggestion
        weblog-variant: ['dd-lib-python-init-test-django','dd-lib-python-init-test-django-gunicorn','dd-lib-python-init-test-django-uvicorn']
```
nit

Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:64]:
```suggestion
        variant: [ 'dd-lib-python-init-test-django','dd-lib-python-init-test-django-gunicorn','dd-lib-python-init-test-django-uvicorn']
```

File: .github/workflows/lib-injection.yml [hunk 8]
```diff
@@ -26,66 +26,56 @@ jobs:
...
+      fail-fast: false
+    env:
+      TEST_LIBRARY: python
+      WEBLOG_VARIANT: ${{ matrix.weblog-variant }}
```
Comment (Kyle-Verhoog) [inline .github/workflows/lib-injection.yml:68]:
```suggestion
      VARIANT: ${{ matrix.variant }}
```

File: lib-injection/Dockerfile [hunk 9]
```diff
@@ -1,3 +1,6 @@
+FROM scratch AS python_agent
```
Comment (Kyle-Verhoog) [inline lib-injection/Dockerfile:1]: can we revert these changes?

File: lib-injection/copy-lib.sh [hunk 10]
```diff
@@ -3,3 +3,6 @@
 # This script is used by the admission controller to install the library from the
 # init container into the application container.
 cp sitecustomize.py "$1/sitecustomize.py"
+if [ -e "dd-trace-py" ]; then
```
Comment (Kyle-Verhoog) [inline lib-injection/copy-lib.sh:6]: here too

File: lib-injection/sitecustomize.py [hunk 11]
```diff
@@ -33,7 +33,7 @@ def _configure_ddtrace():
     except ImportError:
```
Comment (Kyle-Verhoog) [inline lib-injection/sitecustomize.py:33]: revert the changes in this file too

**Review by Kyle-Verhoog — CHANGES_REQUESTED** [review]
:+1: the testing changes look great, thank you!

I do not like the changes to the image building and publishing, as well as the changes made to the code itself to bend itself to the testing framework.

If we'd like to share a building system then I think we should:

1) Propose doing so in a different PR
2) Share the code somewhere other than system-tests
3) Avoid having to change the implementation to work around the tests

File: .github/workflows/lib-injection.yml (hunk 4 above)
Comment (robertomonteromiguel) [reply .github/workflows/lib-injection.yml:29]: Ok, Kyle. I agree with you that system-tests shouldn't take part on the "init-image" generation, i think it is a duty of each team and it is important keep in mind, that these images become public releases.


We have only intervened in two things:

- In dd-trace-java they used shell scripts to create the images, and we changed it to GitHub Actions, similar to how it was in python.
- To run the tests in system-tests we need that the images uploaded to the docker registry (docker hub or Github registry) are always taged the same way. For production we were already using "LATEST". We have added that in the case of the build of the main branch, create a tag "LATEST_SNAPSHOT" (see step lib-injection-tags). Probably the shared Github action that calculates the tags to generate, should be in another repository different from system-tests, but always thinking that we should share this action among all tracer repositories. We can move it later.

In the case of python we did modify the Docker file because when we started with this task, the python image was always being created with the ddTrace release, not with the snapshot version of the branch. 

I see that with the changes you have made this week, I would now take the built snapshot version. Is this correct? @Kyle-Verhoog 

Sorry. I should have asked before modifying it.... Now I'm going to undo the changes in the DockerFile and sitecustomize

File: lib-injection/Dockerfile (hunk 9 above)
Comment (robertomonteromiguel) [reply lib-injection/Dockerfile:1]: Done!

File: lib-injection/sitecustomize.py (hunk 11 above)
Comment (robertomonteromiguel) [reply lib-injection/sitecustomize.py:33]: Done!

File: lib-injection/copy-lib.sh (hunk 10 above)
Comment (robertomonteromiguel) [reply lib-injection/copy-lib.sh:6]: Done!

File: .github/workflows/lib-injection.yml (hunk 7 above)
Comment (robertomonteromiguel) [reply .github/workflows/lib-injection.yml]: I would like to maintain this name, only because in system-tests we called as "weblog". In the future, I would like to use existing "weblog" variants in system-tests, instead to have custom sample apps for lib-injection

File: .github/workflows/lib-injection.yml (hunk 8 above)
Comment (robertomonteromiguel) [reply .github/workflows/lib-injection.yml:63]: I would like to maintain this name, only because in system-tests we called as "weblog". In the future, I would like to use existing "weblog" variants in system-tests, instead to have custom sample apps for lib-injection

File: lib-injection/sitecustomize.py [hunk 12]
```diff
@@ -54,4 +54,4 @@ def _configure_ddtrace():
             _configure_ddtrace()
     else:
         print("datadog autoinstrumentation: ddtrace already installed, skipping install")
-        _configure_ddtrace()
+        _configure_ddtrace()
```
Comment (Kyle-Verhoog) [inline lib-injection/sitecustomize.py:57]:
```suggestion
        _configure_ddtrace()

```

File: lib-injection/copy-lib.sh [hunk 13]
```diff
@@ -2,4 +2,4 @@
 
 # This script is used by the admission controller to install the library from the
 # init container into the application container.
-cp sitecustomize.py "$1/sitecustomize.py"
+cp sitecustomize.py "$1/sitecustomize.py"
```
Comment (Kyle-Verhoog) [inline lib-injection/copy-lib.sh:5]:
```suggestion
cp sitecustomize.py "$1/sitecustomize.py"

```

File: lib-injection/Dockerfile [hunk 14]
```diff
@@ -12,4 +12,4 @@ WORKDIR /datadog-init
 ADD sitecustomize.py /datadog-init/sitecustomize.py
 # Use ~ as a delimiter because git urls can contain slashes
 RUN sed -i "s~<DD_TRACE_VERSION_TO_BE_REPLACED>~${DDTRACE_PYTHON_VERSION}~g" /datadog-init/sitecustomize.py
-ADD copy-lib.sh /datadog-init/copy-lib.sh
+ADD copy-lib.sh /datadog-init/copy-lib.sh
```
Comment (Kyle-Verhoog) [inline lib-injection/Dockerfile:15]:
```suggestion
ADD copy-lib.sh /datadog-init/copy-lib.sh

```

**Review by Kyle-Verhoog — COMMENTED** [review]
:+1: @robertomonteromiguel why do the jobs take considerably longer now? Looks like they take 10+ minutes now compared to ~5 minutes before (including building, pushing and then running test).

**Comment by robertomonteromiguel** [comment]
> 👍 @robertomonteromiguel why do the jobs take considerably longer now? Looks like they take 10+ minutes now compared to ~5 minutes before (including building, pushing and then running test).

Thanks @Kyle-Verhoog . I didn't realize that I was generating the sample image for all platforms. I have changed it and now the times are similar to the main branch.

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by pawelchcki — APPROVED** [review]
